### PR TITLE
feat/upcoming bill api

### PR DESCRIPTION
**Pull request by gitwilliam**
Contributes to [138361969](https://app.teamgantt.com/boards/82cab2e8-b840-4ce4-95d4-da03bb55b8a2/edit/138361969)

### CODE REVIEW

File: src/Model/PaymentMethod.php [hunk 1]
```diff
@@ -90,6 +96,30 @@ public function setLast4(string $last4): self
...
+    {
+        $this->name = $name;
+
+        return $this;
+    }
+
+    public function getType(): string
+    {
+        return $this->type;
+    }
```
Comment (johncorrelli) [inline src/Model/PaymentMethod.php]: Would/should this be an enum? or are there too many variations coming from BT?

File: src/Model/PaymentMethod.php [hunk 2]
```diff
@@ -90,6 +96,30 @@ public function setLast4(string $last4): self
...
+
+    public function setType(string $type): self
```
Comment (johncorrelli) [inline src/Model/PaymentMethod.php:110]: ~I guess same enum question here...~

~EDIT: the other comment was on the other type.... should we make this an enum~

EDIT 2: I checked the docs & they are just returning it as a string... so :nothing-to-do-here:

File: src/Model/PaymentMethod.php [hunk 3]
```diff
@@ -52,12 +58,11 @@ public function setBillingAddress(Address $address): self
...
-        if (!isset($this->billingAddress)) {
-            return [];
-        }
-
+        // If the last4, name, and type are added to this, then they
+        // will need to be stripped off prior to it being submitted to
+        // Braintree as part of a subscription in PaymentMethodMapper::toRequest.
         return [
-            'billingAddress' => $this->billingAddress->toArray(),
+            'billingAddress' => $this->billingAddress?->toArray(),
```
Comment (johncorrelli) [inline src/Model/PaymentMethod.php:65]: this is a slight behavioral change... where before an unset billing address would return `[]` but now it would return `['billingAddress' => null'];`

Should we revert the change, but keep the comment?

**Review by johncorrelli — CHANGES_REQUESTED** [review]
Just a small change. Nice work! Welcome to DUES!

File: src/Processor/Braintree/Mapper/PaymentMethodMapper.php [hunk 4]
```diff
@@ -77,6 +77,10 @@ public function fromResult($paymentMethod): Token
             $token->setBillingAddress($address);
         }
 
+        if (isset($paymentMethod->last4)) {
+            $token->setLast4($paymentMethod->last4);
+        }
```
[the remark this answers is not in the document]
Comment (gitwilliam) [reply src/Processor/Braintree/Mapper/PaymentMethodMapper.php]: @johncorrelli - I'm probably just misunderstanding something, but this is to get the credit card's last 4 numbers to our API and subsequently up to the web client so we can display them. Perhaps I'm doing it in the wrong place? The intention here was to get the last4 from BT, not send it to the BT.

Are you suggesting that I use a different BT API to get the `last4` from the `nonce`?

image

[the remark this answers is not in the document]
Comment (johncorrelli) [reply src/Processor/Braintree/Mapper/PaymentMethodMapper.php]: Ah! you are right... I read the GH line collapsing wrong... this is good. :nothing-to-do-here:

File: src/Model/PaymentMethod.php [hunk 5]
```diff
@@ -58,6 +61,8 @@ public function toArray(): array
 
         return [
             'billingAddress' => $this->billingAddress->toArray(),
+            'expirationDate' => $this->expirationDate ? $this->expirationDate->format('Y-m-d') : null,
```
[the remark this answers is not in the document]
Comment (gitwilliam) [reply src/Model/PaymentMethod.php]: The final commit doesn't have this in `toArray` anymore, so I think we are good.

File: src/Model/PaymentMethod.php (hunk 3 above)
Comment (gitwilliam) [reply src/Model/PaymentMethod.php]: Yes, I had removed that when I was originally packing all the fields in the `toArray`, but I've reverted it back to what it was originally since I am not doing that anymore.  Thanks for catching that. ❤️

**Review by johncorrelli — APPROVED** [review]
